app/services/bill-runs/two-part-tariff/process-two-part-tariff-returns.service.js

```diff
@@ -0,0 +1,66 @@
+'use strict'
+
+/**
+ * Processes the licences for a two-part tariff bill run for the given billing periods
+ * @module ProcessTwoPartTariffReturnsService
+ */
+
+const BillRunModel = require('../../../models/bill-run.model.js')
+const { calculateAndLogTimeTaken, currentTimeInNanoseconds } = require('../../../lib/general.lib.js')
+const HandleErroredBillRunService = require('../handle-errored-bill-run.service.js')
+const MatchAndAllocateService = require('./match-and-allocate.service.js')
+
+/**
+ * Matches and allocates licences to returns for a two-part tariff bill run for the given billing periods. The results
+ * of this matching process are then persisted to the database ready for the results to be reviewed. The bill run status
+ * is then updated based on whether any licences were matched, or if the process has errored.
+ *
+ * @param {module:BillRunModel} billRun
+ * @param {Object[]} billingPeriods An array of billing periods each containing a `startDate` and `endDate`. For 2PT
+ * this will only ever contain a single period
+ */
+async function go (billRun, billingPeriods) {
+  const { id: billRunId } = billRun
+
+  try {
+    const startTime = currentTimeInNanoseconds()
+
+    await _updateStatus(billRunId, 'processing')
+
+    // `isPopulated` will be set to true if `MatchAndAllocateService` matches at least one licence
+    const isPopulated = await MatchAndAllocateService.go(billRun, billingPeriods)
+
+    await _setBillRunStatus(billRunId, isPopulated)
+
+    calculateAndLogTimeTaken(startTime, 'Process bill run complete', { billRunId, type: 'two_part_tariff' })
+  } catch (error) {
+    await HandleErroredBillRunService.go(billRunId)
+    _logError(billRun, error)
+  }
+}
+
+async function _setBillRunStatus (billRunId, isPopulated) {
+  // If there are no bill licences then the bill run is considered empty. We just need to set the status to indicate
+  // this in the UI
+  if (!isPopulated) {
+    await _updateStatus(billRunId, 'empty')
+    return
+  }
+
+  // If licences are successfully matched to returns then the bill run status is set to 'review'
+  await _updateStatus(billRunId, 'review')
+}
+
+function _logError (billRun, error) {
+  global.GlobalNotifier.omfg('Bill run process errored', { billRun }, error)
+}
+
+async function _updateStatus (billRunId, status) {
+  await BillRunModel.query()
+    .findById(billRunId)
+    .patch({ status })
+}
```

Comment:
A few things are going on here. Some corrections, some suggestions and some come from have recently gone through the billing engine for a second time due to my work on annual.

I've made a bunch of suggested changes below. But I'll list the reasons why here

- Before sending the billing periods to the `MatchAndAllocateService`, we know we only care about `billingPeriods[0]`. Whether we do it here or in `StartBillRunProcessService` was a toss-up. I went for here because then the start service remains consistent across the bill runs and can remain ignorant; it is up to each bill run processing service to decide what to do with the billing periods generated.
- Avoiding `isThing`. IMHO I don't think booleans need to be prefixed with `is` which I know comes from my ruby days. The interweb comes down more on the side of it is fine. But I am 👑 here! (I'm also aware this was already here and I've let it get by me during the supplementary days 😬 )
- Creating too many methods. I do believe in a named method can help explain what a single line of code is doing. But on review, I think we took it too far during supplementary billing. Hence I've gotten rid of `_logError()`
- Cleaned up `_setBillRunStatus()` (I blame too much time with the legacy code during supplementary billing!)
- Remove unnecessary await in `_updateStatus()` (Again, old code we should have known better)
- Update JSDoc comments to match team conventions

```suggestion
/**
 * Matches and allocates licences to returns for a two-part tariff bill run
 *
 * The results of the matching process are then persisted to the database ready for the results to be reviewed. The bill
 * run status is also updated to 'review'.
 *
 * In the unlikely event of no licences match to returns it will set the status to 'empty'. It will also handle updating
 * the bill run if an error occurs during the process.
 *
 * @param {module:BillRunModel} billRun - The two-part tariff bill run being processed
 * @param {Object[]} billingPeriods - An array of billing periods each containing a `startDate` and `endDate`. For 2PT
 * this will only ever contain a single period
 */
async function go (billRun, billingPeriods) {
  const { id: billRunId } = billRun
  // NOTE: billingPeriods come from `DetermineBillingPeriodsService` which always returns an array because it is used by
  // all billing types. For two-part tariff we know it will only contain one because 2PT bill runs are only for a single
  // financial year
  const billingPeriod = billingPeriods[0]

  try {
    const startTime = currentTimeInNanoseconds()

    await _updateStatus(billRunId, 'processing')

    // `populated` will be set to true if `MatchAndAllocateService` processes at least one licence
    const populated = await MatchAndAllocateService.go(billRun, billingPeriod)

    await _setBillRunStatus(billRunId, populated)

    calculateAndLogTimeTaken(startTime, 'Process bill run complete', { billRunId, type: 'two_part_tariff' })
  } catch (error) {
    await HandleErroredBillRunService.go(billRunId)
    global.GlobalNotifier.omfg('Bill run process errored', { billRun }, error)
  }
}

async function _setBillRunStatus (billRunId, populated) {
  // It is highly unlikely no licences were matched to returns. So we default status to 'review'
  let status = 'review'

  // Just in case no licences were found to be matched to returns we set the status to 'empty'
  if (!populated) {
    status = 'empty'
  }

  // Update the bill run's status
  return _updateStatus(billRunId, status)
}

async function _updateStatus (billRunId, status) {
  return BillRunModel.query()
    .findById(billRunId)
    .patch({ status })
}
```